pkg/loader/load.go

```diff
@@ -624,6 +629,82 @@ func (b *batchManager) put(txn *Txn) error {
 	return nil
 }
 
+type txnManager struct {
+	cacheChan    chan *Txn
+	shutdown     chan struct{}
+	cachedSize   int
+	maxCacheSize int
+	cond         *sync.Cond
+	closed       bool
+}
+
+func newTxnManager(maxCacheSize int) *txnManager {
+	return &txnManager{
+		cacheChan:    make(chan *Txn, 1024),
+		maxCacheSize: maxCacheSize,
+		cond:         sync.NewCond(new(sync.Mutex)),
+		shutdown:     make(chan struct{}),
+	}
+}
+
+func (t *txnManager) put(input chan *Txn) chan *Txn {
+	ret := t.cacheChan
+	go func() {
+		for !t.closed {
+			var txn *Txn
+			var ok bool
+			select {
+			case txn, ok = <-input:
+				if !ok {
+					close(ret)
+					return
+				}
+			case <-t.shutdown:
+				return
+			}
+			txnSize := len(txn.DMLs)
+
+			t.cond.L.Lock()
+			if txnSize < t.maxCacheSize {
+				for !t.closed && txnSize+t.cachedSize > t.maxCacheSize {
+					t.cond.Wait()
+				}
+			} else {
+				for !t.closed && t.cachedSize != 0 {
+					t.cond.Wait()
+				}
+			}
+			t.cond.L.Unlock()
+
+			select {
+			case ret <- txn:
+				t.cond.L.Lock()
+				t.cachedSize += txnSize
+				t.cond.L.Unlock()
+			case <-t.shutdown:
+				return
+			}
+		}
+	}()
+	return ret
+}
+
+func (t *txnManager) pop(txn *Txn) {
+	t.cond.L.Lock()
+	t.cachedSize -= len(txn.DMLs)
+	t.cond.Signal()
+	t.cond.L.Unlock()
+}
+
+func (t *txnManager) Close() {
+	if t.closed {
```

Comment:
should we use atomic operation？

Reply:
There won't be concurrent operations to change `t.closed` so I think we don't need to use atomic operation.

Reply:
YOU NEED TO ADD A COMMENT `IT'S NOT THREAD SAFE`